File: client/java/src/main/java/com/be/mesh/client/mpc/MeshEden.java
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */
package com.be.mesh.client.mpc;

import com.be.mesh.client.annotate.*;
import com.be.mesh.client.prsim.*;
import com.be.mesh.client.struct.*;
import com.be.mesh.client.tool.LDCable;
import com.be.mesh.client.tool.Mode;
import com.be.mesh.client.tool.Tool;
import lombok.Getter;
import lombok.Setter;
import lombok.extern.slf4j.Slf4j;

import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.lang.reflect.Parameter;
import java.lang.reflect.Type;
import java.time.Duration;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * Like: create.tenant.omega.json.http2.lx000001.mpi.trustbe.cn
 *
 * @author coyzeng@gmail.com
 */
@Slf4j
@SPI("mesh")
@Listener(topic = MeshEden.TOPIC, code = MeshEden.CODE)
public class MeshEden implements Eden, RuntimeHook, Subscriber {

    private String aliveId = "";
    public static final String TOPIC = "mesh.registry.event";
    public static final String CODE = "refresh";
    public static final Duration DEFAULT_TIMEOUT = Duration.ofSeconds(10);
    private final Map<Class<?>, Provider> providers = new ConcurrentHashMap<>();
    private final Map<Class<?>, Map<MPI, Consumer>> consumers = new ConcurrentHashMap<>();
    private final Map<Class<?>, Map<MPI, Map<String, Instance<Reference>>>> indies = new ConcurrentHashMap<>();
    private final Map<String, Instance<Service>> services = new ConcurrentHashMap<>();
    private final Map<String, Instance<Reference>> references = new ConcurrentHashMap<>();

    @Override
    public Object define(MPI mpi, Class<?> reference) {
        return makeConsumer(mpi, reference).getProxy();
    }

    @Override
    public Execution<Reference> refer(MPI mpi, Class<?> reference, Inspector inspector) {
        return makeReferenceExecution(mpi, reference, getEnviron()).get(inspector.getSignature());
    }

    @Override
    public void store(Class<?> type, Object service) {
        providers.computeIfAbsent(type, key -> new Provider(type, service));
    }

    @Override
    public Execution<Service> infer(String urn) {
        URN domain = URN.from(urn);
        Map<String, Instance<Service>> executions = makeServiceExecution(getEnviron());
        return Optional.ofNullable(executions.get(domain.getName())).orElseGet(() -> {
            for (Map.Entry<String, Instance<Service>> instance : executions.entrySet()) {
                if (domain.matchName(instance.getKey())) {
                    return instance.getValue();
                }
            }
            return null;
        });
    }

    @Override
    public List<Class<?>> referTypes() {
        return new ArrayList<>(consumers.keySet());
    }

    @Override
    public List<Class<?>> inferTypes() {
        return providers.values().stream().map(Provider::getKind).collect(Collectors.toList());
    }

    private Environ getEnviron() {
        Network network = ServiceLoader.load(Network.class).getDefault();
        return network.getEnviron();
    }

    private Consumer makeConsumer(MPI mpi, Class<?> reference) {
        return consumers.computeIfAbsent(reference, key -> new ConcurrentHashMap<>()).computeIfAbsent(mpi, key -> new Consumer(mpi, reference));
    }

    private Map<String, Instance<Reference>> makeReferenceExecution(MPI mpi, Class<?> reference, Environ environ) {
        Consumer consumer = makeConsumer(mpi, reference);
        return indies.computeIfAbsent(reference, key -> new ConcurrentHashMap<>()).computeIfAbsent(mpi, key -> {
            Map<String, Instance<Reference>> executions = new ConcurrentHashMap<>();
            getDeclaredMethods(reference).forEach((kind, methods) -> methods.forEach(method -> {
                StringBuilder signature = new StringBuilder(method.getName());
                for (Parameter parameter : method.getParameters()) {
                    signature.append(parameter.getType().getName());
                }
                Reference refer = makeMethodAsReference(environ, mpi, kind, method);
                Instance<Reference> instance = new Instance<>(refer.getUrn(), reference, method, consumer.getProxy(), refer);
                references.put(refer.getUrn(), instance);
                executions.put(signature.toString(), instance);
            }));
            return executions;
        });
    }

    private Map<String, Instance<Service>> makeServiceExecution(Environ environ) {
        providers.values().stream().filter(x -> !x.isMaking()).forEach(provider -> {
            Class<?> type = provider.getKind();
            Object service = provider.getService();
            getDeclaredMethods(type).forEach((kind, methods) -> methods.forEach(method -> {
                List<Service> schemas = makeSchema(environ, type, kind, method);
                for (Service schema : schemas) {
                    services.put(URN.from(schema.getUrn()).getName(), new Instance<>(schema.getUrn(), type, method, service, schema));
                }
            }));
            provider.setMaking(true);
        });
        return services;
    }

    /**
     * Build service schema.
     *
     * @param environ Environ
     * @param type    Class type.
     * @param kind    Interface type.
     * @param method  Invoked method.
     * @return Schemas
     */
    private List<Service> makeSchema(Environ environ, Class<?> type, Class<?> kind, Method method) {
        List<Service> schemas = new ArrayList<>();
        MPS mps = type.getAnnotation(MPS.class);
        if (null != mps) {
            Service service = makeMethodAsService(mps, kind, method, environ);
            if (!Tool.MESH_MODE.get().match(Mode.NoCommunal) || (service.getFlags() & 4) != 4) {
                schemas.add(service);
            }
        }
        if (Subscriber.class != kind) {
            return schemas;
        }
        Bindings bindings = type.getAnnotation(Bindings.class);
        if (null != bindings) {
            for (Binding binding : bindings.value()) {
                if (binding.meshable()) {
                    schemas.add(makeMethodAsService(binding, kind, method, environ));
                }
            }
        }
        Binding binding = type.getAnnotation(Binding.class);
        if (null != binding && binding.meshable()) {
            schemas.add(makeMethodAsService(binding, kind, method, environ));
        }
        return schemas;
    }

    private <T> T chooseIfPresent(Function<MPI, T> fn, T defaults, MPI... tags) {
        for (MPI tag : tags) {
            if (null != tag && Tool.required(fn.apply(tag)) && !Tool.equals(fn.apply(tag), defaults)) {
                return fn.apply(tag);
            }
        }
        return defaults;
    }

    private <T> Stream<T> choose(Function<MPI, T> fn, T defaults, MPI... tags) {
        List<T> vs = new ArrayList<>();
        for (MPI tag : tags) {
            if (null != tag && Tool.required(fn.apply(tag)) && !Tool.equals(fn.apply(tag), defaults)) {
                vs.add(fn.apply(tag));
            }
        }
        vs.add(defaults);
        return vs.stream();
    }

    private Reference makeMethodAsReference(Environ environ, MPI mpi, Class<?> type, Method method) {
        MPI om = method.getAnnotation(MPI.class);
        MPI ok = type.getAnnotation(MPI.class);
        String value = chooseIfPresent(MPI::value, "", om);
        String name = chooseIfPresent(MPI::name, "", om);
        String proto = chooseIfPresent(MPI::proto, MeshFlag.GRPC.getName(), mpi, om, ok);
        String codec = chooseIfPresent(MPI::codec, MeshFlag.JSON.getName(), mpi, om, ok);
        String alias = Tool.anyone(value, name, Tool.toLowerCase(String.format("%s.%s", type.getName(), method.getName())));
        Reference reference = new Reference();
        reference.setNamespace(method.getDeclaringClass().getName());
        reference.setName(method.getName());
        reference.setVersion(chooseIfPresent(MPI::version, "", mpi, om, ok));
        reference.setProto(Tool.anyone(proto, MeshFlag.GRPC.getName()));
        reference.setCodec(Tool.anyone(codec, MeshFlag.JSON.getName()));
        reference.setFlags(chooseIfPresent(MPI::flags, 0L, om, ok) | mpi.flags());
        reference.setTimeout(choose(MPI::timeout, DEFAULT_TIMEOUT.toMillis(), mpi, om, ok).max(Long::compare).orElse(DEFAULT_TIMEOUT.toMillis()));
        reference.setRetries(choose(MPI::retries, 3, mpi, om, ok).max(Integer::compare).orElse(3));
        reference.setNode(chooseIfPresent(MPI::node, "", mpi, om, ok));
        reference.setInst(chooseIfPresent(MPI::inst, "", mpi, om, ok));
        reference.setZone(chooseIfPresent(MPI::zone, Tool.MESH_ZONE.get(), mpi, om, ok));
        reference.setCluster(chooseIfPresent(MPI::cluster, Tool.MESH_CLUSTER.get(), mpi, om, ok));
        reference.setCell(chooseIfPresent(MPI::cell, Tool.MESH_CELL.get(), mpi, om, ok));
        reference.setGroup(chooseIfPresent(MPI::group, Tool.MESH_GROUP.get(), mpi, om, ok));
        reference.setAddress(chooseIfPresent(MPI::address, "", mpi, om, ok));
        reference.setUrn(getURN(alias, reference, environ));
        return reference;
    }

    private Service makeMethodAsService(MPS mps, Class<?> kind, Method method, Environ environ) {
        MPS md = Optional.ofNullable(method.getAnnotation(MPS.class)).orElse(mps);
        MPI om = method.getAnnotation(MPI.class);
        MPI ok = kind.getAnnotation(MPI.class);
        String value = chooseIfPresent(MPI::value, "", om).replace("${mesh.name}", Tool.MESH_NAME.get()).replace("${mesh.uname}", md.name());
        String name = chooseIfPresent(MPI::name, "", om).replace("${mesh.name}", Tool.MESH_NAME.get()).replace("${mesh.uname}", md.name());
        String proto = chooseIfPresent(MPI::proto, Tool.anyone(md.proto(), MeshFlag.GRPC.getName()), om, ok);
        String codec = chooseIfPresent(MPI::codec, Tool.anyone(md.codec(), MeshFlag.JSON.getName()), om, ok);
        String alias = Tool.anyone(value, name, Tool.toLowerCase(String.format("%s.%s", kind.getName(), method.getName())));
        Service service = initServiceEnviron(environ);
        service.setNamespace(method.getDeclaringClass().getName());
        service.setName(md.name());
        service.setVersion(md.version());
        service.setProto(proto);
        service.setCodec(codec);
        service.setFlags(chooseIfPresent(MPI::flags, 0L, om, ok) | md.flags());
        service.setTimeout(choose(MPI::timeout, md.timeout(), om, ok).max(Long::compare).orElse(DEFAULT_TIMEOUT.toMillis()));
        service.setKind("MPS");
        service.setUrn(getURN(alias, service, environ));
        return service;
    }

    private Service initServiceEnviron(Environ environ) {
        Service service = new Service();
        service.setRetries(0);
        service.setNode(environ.getNodeId());
        service.setInst(environ.getInstId());
        service.setZone(Optional.ofNullable(environ.getDistribution()).map(Distribution::getZone).orElse(Tool.MESH_ZONE.get()));
        service.setCluster(Optional.ofNullable(environ.getDistribution()).map(Distribution::getCluster).orElse(Tool.MESH_CLUSTER.get()));
        service.setCell(Optional.ofNullable(environ.getDistribution()).map(Distribution::getCell).orElse(Tool.MESH_CELL.get()));
        service.setGroup(Optional.ofNullable(environ.getDistribution()).map(Distribution::getGroup).orElse(Tool.MESH_GROUP.get()));
        service.setSets(Tool.MESH_NAME.get());
        service.setAddress(getListenedAddress());
        service.setLang("Java");
        service.setAttrs(new HashMap<>(0));
        return service;
    }

    private Service makeMethodAsService(Binding binding, Class<?> kind, Method method, Environ environ) {
        Binding metadata = Optional.ofNullable(method.getAnnotation(Binding.class)).orElse(binding);
        String proto = Tool.anyone(metadata.proto(), MeshFlag.GRPC.getName());
        String codec = Tool.anyone(metadata.codec(), MeshFlag.JSON.getName());
        String alias = Tool.anyone(Tool.toLowerCase(String.format("%s.%s", metadata.topic(), metadata.code())), Tool.toLowerCase(String.format("%s.%s", kind.getName(), method.getName())));
        Service service = initServiceEnviron(environ);
        service.setNamespace(metadata.topic());
        service.setName(metadata.code());
        service.setVersion(metadata.version());
        service.setProto(proto);
        service.setCodec(codec);
        service.setFlags(metadata.flags());
        service.setTimeout(metadata.timeout() < 10 ? DEFAULT_TIMEOUT.toMillis() : metadata.timeout());
        service.setKind("Binding");
        service.setUrn(getURN(alias, service, environ));
        return service;
    }

    private Map<Class<?>, List<Method>> getDeclaredMethods(Class<?> type) {
        List<Class<?>> kinds = new ArrayList<>(type.getInterfaces().length + 1);
        kinds.addAll(Arrays.asList(type.getInterfaces()));
        if (type.isInterface()) {
            kinds.add(type);
        }
        return kinds.stream().collect(Collectors.toMap(kind -> kind, kind -> Arrays.stream(kind.getDeclaredMethods()).filter(method -> method.getDeclaringClass() != Object.class && Modifier.isPublic(method.getModifiers()) && !Modifier.isStatic(method.getModifiers())).collect(Collectors.toList())));
    }

    /**
     * Get the definition or reference definition.
     *
     * @param alias      Service alias name.
     * @param definition URN flags.
     * @param environ    Net environ
     * @return Unique uniform resource domain name.
     */
    private String getURN(String alias, LDCable definition, Environ environ) {
        URN urn = new URN();
        urn.setDomain(URN.MESH_DOMAIN);
        urn.setNodeId(Tool.toLowerCase(Tool.anyone(definition.getNode(), definition.getInst(), environ.getNodeId())));
        urn.setName(alias);
        urn.setFlag(getURNFlag(definition));
        return urn.toString();
    }

    /**
     * Get the reference or reference definition.
     *
     * @param definition Service definition.
     * @return Unique uniform resource domain name.
     */
    private URNFlag getURNFlag(LDCable definition) {
        String[] authority = Tool.split(definition.getAddress(), ":");
        URNFlag flag = new URNFlag();
        flag.setV("00");
        flag.setProto(MeshFlag.ofName(definition.getProto()).getCode());
        flag.setCodec(MeshFlag.ofName(definition.getCodec()).getCode());
        flag.setVersion(definition.getVersion());
        flag.setZone(definition.getZone());
        flag.setCluster(definition.getCluster());
        flag.setCell(definition.getCell());
        flag.setGroup(definition.getGroup());
        flag.setAddress(authority.length > 0 ? authority[0] : "");
        flag.setPort(authority.length > 1 ? authority[1] : "");
        return flag;
    }

    private String getListenedAddress() {
        // For ICBC PaaS
        String xh = Tool.MESH_RUNTIME_IP.get();
        String xp = Tool.MESH_RUNTIME_PORT.get();
        if (Tool.required(xh) && Tool.required(xp)) {
            return String.format("%s:%s", xh, xp);
        }
        return String.format("%s:%d", Tool.MESH_RUNTIME.get().getHost(), Tool.MESH_RUNTIME.get().getPort());
    }

    private Registration newRegistration(Metadata metadata, Map<String, String> attachments) {
        Registration registration = new Registration();
        registration.setInstanceId(getListenedAddress());
        registration.setContent(metadata);
        registration.setKind(Registration.METADATA);
        registration.setAddress(getListenedAddress());
        registration.setName(Tool.MESH_NAME.get());
        registration.setTimestamp(System.currentTimeMillis());
        registration.setAttachments(attachments);
        return registration;
    }

    @Override
    public void stop() throws Throwable {
        Scheduler scheduler = ServiceLoader.load(Scheduler.class).getDefault();
        scheduler.cancel(aliveId);
        scheduler.shutdown(Duration.ofSeconds(3));
        Registry registry = ServiceLoader.load(Registry.class).getDefault();
        Metadata metadata = new Metadata();
        metadata.setReferences(Collections.emptyList());
        metadata.setServices(Collections.emptyList());
        Tool.MESH_ADDRESS.get().getServers().parallelStream().forEach(addr -> Mesh.contextSafeUncheck(() -> {
            try {
                Mesh.context().setAttribute(Mesh.REMOTE, addr.getAddress());
                registry.unregister(newRegistration(metadata, new HashMap<>(0)));
            } catch (Exception e) {
                log.warn("Register to {} with {}:{}", addr.getAddress(), e.getClass().getName(), e.getMessage());
            }
        }));
    }

    @Override
    public void refresh() throws Throwable {
        start();
        Scheduler scheduler = ServiceLoader.load(Scheduler.class).getDefault();
        aliveId = scheduler.period(Duration.ofSeconds(30), new Topic(MeshEden.TOPIC, MeshEden.CODE));
    }

    @Override
    public void subscribe(Event event) {
        Builtin builtin = ServiceLoader.load(Builtin.class).getDefault();
        Versions versions = Optional.ofNullable(builtin.version()).orElseGet(Versions::new);
        Map<String, String> attachments = new HashMap<>();
        attachments.put("_PAAS_PORT_7700", Tool.getProperty("", "_PAAS_PORT_7700"));
        attachments.putAll(Optional.ofNullable(versions.getInfos()).orElseGet(Collections::emptyMap));
        Context.Metadata.MESH_SUBSET.set(attachments, Tool.MESH_SUBSET.get());
        Environ environ = getEnviron();
        makeServiceExecution(environ);
        consumers.values().stream().flatMap(x -> x.values().stream()).forEach(consumer -> makeReferenceExecution(consumer.getMpi(), consumer.getReference(), environ));
        Registry registry = ServiceLoader.load(Registry.class).getDefault();
        Metadata metadata = new Metadata();
        metadata.setReferences(references.values().stream().map(Instance::getResource).collect(Collectors.toList()));
        metadata.setServices(services.values().stream().map(Instance::getResource).collect(Collectors.toList()));

        Tool.MESH_ADDRESS.get().getServers().parallelStream().forEach(addr -> Mesh.contextSafeUncheck(() -> {
            try {
                Mesh.context().setAttribute(Mesh.REMOTE, addr.getAddress());
                registry.register(newRegistration(metadata, attachments));
            } catch (Exception e) {
                log.warn("Register to {} with {}:{}", addr.getAddress(), e.getClass().getName(), e.getMessage());
            }
        }));
    }

    @Getter
    private static final class Instance<T> implements Execution<T> {

        private final String urn;
        private final Class<?> type;
        private final Method method;
        private final Object target;
        private final T resource;
        private final Inspector inspector;
        private final Invoker<?> invoker;
        private final Class<? extends Parameters> intype;
        private final Class<? extends Returns> retype;

        private Instance(String urn, Class<?> type, Method method, Object target, T resource) {
            this.urn = urn;
            this.type = type;
            this.method = method;
            this.target = target;
            this.resource = resource;
            this.inspector = new MethodInspector(type, method);
            this.invoker = new ServiceInvokeHandler(target);
            this.intype = ServiceLoader.load(JCompiler.class).get(JCompiler.JAVASSIST).intype(this.method);
            this.retype = ServiceLoader.load(JCompiler.class).get(JCompiler.JAVASSIST).retype(this.method);
        }

        @Override
        public T schema() {
            return this.resource;
        }

        @Override
        public Inspector inspect() {
            return this.inspector;
        }

        @Override
        public <I extends Parameters> Types<I> intype() {
            return Types.of((Type) this.intype);
        }

        @Override
        public <O extends Returns> Types<O> retype() {
            return Types.of((Type) this.retype);
        }

        @Override
        public Parameters inflect() {
            return Tool.newInstance(this.intype);
        }

        @Override
        public Returns reflect() {
            return Tool.newInstance(this.retype);
        }

        @Override
        public Object invoke(Invocation invocation) throws Throwable {
            return this.invoker.invoke(invocation);
        }
    }

    @Getter
    private static final class Consumer {
        private final MPI mpi;
        private final Class<?> reference;
        private final Object proxy;

        public Consumer(MPI mpi, Class<?> reference) {
            this.mpi = mpi;
            this.reference = reference;
            this.proxy = ServiceProxy.proxy(mpi, reference);
        }
    }

    @Getter
    private static final class Provider {
        @Setter
        private boolean making;
        private final Class<?> kind;
        private final Object service;

        public Provider(Class<?> kind, Object service) {
            this.kind = kind;
            this.service = service;
        }
    }
}

File: client/python/mesh/mpc/mesh_eden.py
#
# Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
# TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
#
#
import inspect
import os
from typing import Type, Any, Generic, Dict, List, Callable

import mesh.kinds as kinds
import mesh.log as log
import mesh.tool as tool
from mesh.kinds import Reference, Service, Environ, MeshFlag, Event, Registration, Resource, Topic, Versions
from mesh.macro import ark, T, A, mpi, MPI, spi, mps, MPS, binding, Binding, ServiceLoader, Returns, Parameters, \
    Inspector
from mesh.mpc.compiler import Compiler
from mesh.mpc.eden import Eden
from mesh.mpc.inspector import MethodInspector
from mesh.mpc.invoker import Invocation, Execution
from mesh.mpc.service import ServiceInvokeHandler
from mesh.mpc.service_proxy import ServiceProxy
from mesh.mpc.urn import URN, URNFlag, MESH_DOMAIN
from mesh.prsim import Network, RuntimeHook, Subscriber, Registry, Scheduler, Builtin


class Provider:

    def __init__(self, kind: Type[T], service: Any):
        self.making: bool = False
        self.kind: Type[T] = kind
        self.service: Any = service


class Consumer:
    def __init__(self, metadata: mpi, reference: Type[T]):
        self.metadata: mpi = metadata
        self.reference: Type[T] = reference
        self.proxy: Any = ServiceProxy.static_proxy(reference, metadata)


class Instance(Generic[T], Execution[T]):

    def __init__(self, urn: str, kind: Type[T], method: Any, target: Any, resource: T):
        declared_method = getattr(kind, method.__name__)
        self.urn = urn
        self.kind = kind
        self.method = method
        self.target = target
        self.resource = resource
        self.inspector = MethodInspector(MPI.get_mpi_if_present(method), kind, method)
        self.invoker = ServiceInvokeHandler(target)
        self.__intype = ServiceLoader.load(Compiler).get_default().intype(declared_method)
        self.__retype = ServiceLoader.load(Compiler).get_default().retype(declared_method)

    def schema(self) -> T:
        return self.resource

    def inspect(self) -> Inspector:
        return self.inspector

    def intype(self) -> Type[Parameters]:
        return self.__intype

    def retype(self) -> Type[Returns]:
        return self.__retype

    def inflect(self) -> Parameters:
        return self.__intype()

    def reflect(self) -> Returns:
        return self.__retype()

    def run(self, invocation: Invocation) -> Any:
        return self.invoker.run(invocation)


@spi("mesh")
@binding(topic="mesh.registry.event", code="refresh", meshable=False)
class MeshEden(Eden, RuntimeHook, Subscriber):
    """
    Like: create.tenant.omega.json.http2.lx000001.mpi.trustbe.net
    """

    def __init__(self):
        self.topic = Topic()
        self.topic.topic = "mesh.registry.event"
        self.topic.code = "refresh"
        self.task_id = ""
        self.providers: Dict[Type[T], Provider] = {}
        self.consumers: Dict[Type[T], Dict[mpi, Consumer]] = {}
        self.indies: Dict[Type[T], Dict[mpi, Dict[Any, Instance[Reference]]]] = {}
        self.services: Dict[str, Instance[Service]] = {}
        self.references: Dict[str, Instance[Reference]] = {}

    def define(self, metadata: mpi, reference: Type[T]) -> T:
        return self.make_consumer(metadata, reference)

    def refer(self, metadata: mpi, reference: Type[T], method: Inspector) -> Execution[Reference]:
        return self.make_reference_exec(metadata, reference, self.get_environ()).get(method.get_signature())

    def store(self, kind: Type[T], service: Any):
        if not self.providers.get(kind):
            self.providers[kind] = Provider(kind, service)
        return self.providers.get(kind)

    def infer(self, urn: str) -> Execution[Service]:
        return self.make_service_exec(self.get_environ()).get(URN.parse(urn).name)

    def refer_types(self) -> List[Type[T]]:
        return list(self.consumers.keys())

    def infer_types(self) -> List[Type[T]]:
        types: List[Type[T]] = []
        for provider in self.providers.values():
            types.append(provider.kind)
        return types

    def make_consumer(self, metadata: mpi, reference: Type[T]):
        if not self.consumers.get(reference):
            self.consumers[reference] = {}
        if not self.consumers.get(reference).get(metadata):
            self.consumers[reference][metadata] = Consumer(metadata, reference)
        return self.consumers.get(reference).get(metadata)

    def make_reference_exec(self, md: mpi, ref: Type[T], env: Environ) -> Dict[Any, Instance[Reference]]:
        consumer = self.make_consumer(md, ref)
        if not self.indies.get(ref):
            self.indies[ref] = {}
        if not self.indies.get(ref).get(md):
            executions: Dict[Any, Instance[Reference]] = {}
            for kind, methods in tool.get_declared_methods(ref).items():
                for method in methods:
                    refer = self.make_method_as_reference(env, md, kind, method)
                    instance = Instance(refer.urn, ref, method, consumer.proxy, refer)
                    self.references[refer.urn] = instance
                    executions[instance.inspect().get_signature()] = instance
            self.indies.get(ref)[md] = executions
        return self.indies[ref][md]

    def make_service_exec(self, environ: Environ) -> Dict[str, Instance[Service]]:
        for provider in self.providers.values():
            if provider.making:
                continue
            for kind, methods in tool.get_declared_methods(provider.kind).items():
                for method in methods:
                    if not hasattr(provider.service, method.__name__):
                        continue
                    mm = getattr(provider.service, method.__name__)
                    schemas: List[Service] = self.make_schema(environ, provider.kind, kind, method)
                    for schema in schemas:
                        urn = URN.parse(schema.urn)
                        self.services[urn.name] = Instance(schema.urn, provider.kind, mm, provider.service, schema)
            provider.making = True
        return self.services

    @staticmethod
    def get_environ() -> Environ:
        network = ServiceLoader.load(Network).get_default()
        return network.get_environ()

    def make_schema(self, environ: Environ, interface: T, kind: T, method: Any) -> List[Service]:
        schemas = []
        macro = MPS.get_mps_if_present(interface)
        if macro:
            schemas.append(self.make_mps_as_service(macro, kind, method, environ))
        bindings = Binding.get_binding_if_present(interface)
        for metadata in bindings:
            if metadata.meshable:
                schemas = schemas.append(self.make_binding_as_service(metadata, kind, method, environ))
        return schemas

    def make_method_as_reference(self, environ: Environ, metadata: MPI, interface: T, method: Any) -> Reference:
        om = MPI.get_mpi_if_present(method)
        ok = MPI.get_mpi_if_present(interface)
        name = self.choose_if_present(lambda x: x.name, "", om)
        alias = tool.anyone(name, str.lower(f'{interface.__name__}.{method.__name__}'))
        reference = Reference()
        reference.namespace = interface.__name__
        reference.name = method.__name__
        reference.version = self.choose_if_present(lambda x: x.version, '', metadata, om, ok)
        reference.proto = self.choose_if_present(lambda x: x.proto, MeshFlag.GRPC.name, metadata, om, ok)
        reference.codec = self.choose_if_present(lambda x: x.codec, MeshFlag.JSON.name, metadata, om, ok)
        reference.flags = self.choose_if_present(lambda x: x.flags, 0, om, ok) | metadata.flags
        reference.timeout = self.choose(lambda x: x.timeout, 10000, metadata, om, ok)
        reference.retries = self.choose(lambda x: x.retries, 3, metadata, om, ok)
        reference.node = self.choose_if_present(lambda x: x.node, "", metadata, om, ok)
        reference.inst = self.choose_if_present(lambda x: x.inst, "", metadata, om, ok)
        reference.zone = self.choose_if_present(lambda x: x.zone, "", metadata, om, ok)
        reference.cluster = self.choose_if_present(lambda x: x.cluster, "", metadata, om, ok)
        reference.cell = self.choose_if_present(lambda x: x.cell, "", metadata, om, ok)
        reference.group = self.choose_if_present(lambda x: x.group, "", metadata, om, ok)
        reference.address = self.choose_if_present(lambda x: x.address, "", metadata, om, ok)
        reference.urn = self.get_urn(alias, self.get_reference_urn_flag(reference), environ)
        return reference

    def make_mps_as_service(self, metadata: MPS, kind: T, method: Any, environ: Environ) -> Service:
        om = MPI.get_mpi_if_present(method)
        ok = MPI.get_mpi_if_present(kind)
        name = self.choose_if_present(lambda x: x.name, "", om).replace("${mesh.name}", tool.get_mesh_name()). \
            replace("${mesh.uname}", metadata.name)
        alias = tool.anyone(name, str.lower(f'{kind.__name__}.{method.__name__}'))
        service = self.init_service(environ)
        service.kind = kinds.MPS
        service.namespace = kind.__name__
        service.name = method.__name__
        service.proto = self.choose_if_present(lambda x: x.proto, MeshFlag.GRPC.name, metadata, om, ok)
        service.codec = self.choose_if_present(lambda x: x.codec, MeshFlag.JSON.name, metadata, om, ok)
        service.version = self.choose_if_present(lambda x: x.version, metadata.version, metadata, om, ok)
        service.flags = self.choose_if_present(lambda x: x.flags, 0, om, ok) | metadata.flags
        service.timeout = self.choose(lambda x: x.timeout, metadata.timeout, metadata, om, ok)
        service.urn = self.get_urn(alias, self.get_service_urn_flag(service), environ)
        return service

    def make_binding_as_service(self, metadata: Binding, kind: T, method: Any, environ: Environ) -> Service:
        service = self.init_service(environ)
        service.kind = kinds.Binding
        service.namespace = metadata.topic
        service.name = metadata.code
        service.proto = self.choose_if_present(lambda x: x.proto, MeshFlag.GRPC.name, metadata)
        service.codec = self.choose_if_present(lambda x: x.codec, MeshFlag.JSON.name, metadata)
        service.version = metadata.version
        service.flags = metadata.flags
        service.timeout = self.choose(lambda x: x.timeout, metadata.timeout, metadata)
        service.urn = self.get_urn(str.lower(f'{metadata.topic}.{metadata.code}'), self.get_service_urn_flag(service),
                                   environ)
        return service

    def init_service(self, environ: Environ) -> Service:
        service = Service()
        service.namespace = ''
        service.name = ''
        service.version = ''
        service.proto = ''
        service.codec = ''
        service.flags = 0
        service.timeout = 10000
        service.retries = 0
        service.node = environ.node_id
        service.inst = environ.inst_id
        service.zone = environ.lattice.zone
        service.cluster = environ.lattice.cluster
        service.cell = environ.lattice.cell
        service.group = environ.lattice.group
        service.address = self.get_listened_address()
        service.kind = kinds.MPS
        service.lang = 'Python3'
        service.attrs = {}
        service.urn = ''
        return service

    @staticmethod
    def get_urn(alias: str, flag: URNFlag, environ: Environ) -> str:
        urn = URN()
        urn.domain = MESH_DOMAIN
        urn.node_id = str.lower(environ.node_id)
        urn.name = alias
        urn.flag = flag
        return urn.string()

    @staticmethod
    def get_reference_urn_flag(reference: Reference) -> URNFlag:
        authority = str.split(f'{reference.address}:', ':')
        flag = URNFlag()
        return flag.reset(MeshFlag.of_name(reference.proto).get_code(),
                          MeshFlag.of_name(reference.codec).get_code(),
                          reference.version,
                          reference.zone,
                          reference.cluster,
                          reference.cell,
                          reference.group,
                          authority[0],
                          authority[1])

    @staticmethod
    def get_service_urn_flag(service: Service) -> URNFlag:
        flag = URNFlag()
        return flag.reset(MeshFlag.of_name(service.proto).get_code(),
                          MeshFlag.of_name(service.codec).get_code(),
                          service.version,
                          service.zone,
                          service.cluster,
                          service.cell,
                          service.group,
                          '',
                          '')

    @staticmethod
    def choose_if_present(fn: Callable[[A], T], defaults: T, *tags: A) -> T:
        for tag in tags:
            if tag and fn(tag):
                return fn(tag)

        return defaults

    @staticmethod
    def choose(fn: Callable[[A], T], defaults: T, *tags: A) -> T:
        for tag in tags:
            if tag and fn(tag) and fn(tag) != defaults:
                return fn(tag)

        return defaults

    @staticmethod
    def get_listened_address() -> str:
        return f'{tool.get_mesh_runtime().hostname}:{tool.get_mesh_runtime().port}'

    def new_registration(self, metadata: Resource, attachments: Dict[str, str]) -> Registration:
        registration = Registration()
        registration.instance_id = self.get_listened_address()
        registration.content = metadata
        registration.kind = Registration.METADATA
        registration.address = self.get_listened_address()
        registration.name = tool.get_mesh_name()
        registration.timestamp = 30 * 1000
        registration.attachments = attachments
        return registration

    @mpi
    def registry(self) -> Registry:
        """ """
        pass

    @spi("mesh")
    def scheduler(self) -> Scheduler:
        return ServiceLoader.load(Scheduler).get_default()

    @spi("mesh")
    def builtin(self) -> Builtin:
        return ServiceLoader.load(Builtin).get_default()

    def start(self):
        for kind, nss in ark.export(mps).items():
            for spec in ark.providers(mps, kind):
                if not isinstance(spec.metadata, MPS) or inspect.isabstract(spec.kind):
                    continue
                try:
                    self.store(kind, spec.kind())
                    break
                except BaseException as e:
                    log.warn(f"Mesh service {kind.__name__} with {e.__str__()}")

        if os.environ.get("MESH_REGISTER", "0") != "1":
            return

        self.task_id = self.scheduler().period(30 * 1000, self.topic)

    def stop(self):
        self.scheduler().cancel(self.task_id)
        self.scheduler().shutdown(3 * 1000)
        metadata = Resource()
        metadata.services = []
        metadata.references = []
        try:
            self.registry().unregister(self.new_registration(metadata, {}))
        except BaseException as e:
            log.warn("Graceful unregister metadata, {}", str(e))

    def refresh(self):
        self.start()

    def subscribe(self, event: Event):
        v = self.builtin().version()
        version = v if v else Versions()
        attachments = {}
        for k, v in (version.infos if version.infos else {}).items():
            attachments[k] = v
        environ = self.get_environ()
        self.make_service_exec(environ)
        for _, consumers in self.consumers.items():
            for _, consumer in consumers.items():
                self.make_reference_exec(consumer.metadata, consumer.reference, environ)
        metadata = Resource()
        metadata.services = []
        for _, reference in self.references.items():
            metadata.services.append(reference.resource)
        metadata.references = []
        for _, service in self.services.items():
            metadata.references.append(service.resource)
        self.registry().register(self.new_registration(metadata, attachments))
